src/main/java/dinkplugin/notifiers/PetNotifier.java

```diff
@@ -215,56 +236,311 @@ private static class ParseResult {
         boolean collectionLog;
     }
 
+    private static abstract class Source {
+        abstract Double getProbability(Client client, KillCountService kcService);
+
+        abstract Integer estimateActions(Client client, KillCountService kcService);
+
+        Double calculateLuck(Client client, KillCountService kcService, double probability, int killCount) {
+            return MathUtils.cumulativeGeometric(probability, killCount);
+        }
+    }
+
+    @Value
+    @EqualsAndHashCode(callSuper = true)
+    private static class SkillSource extends Source {
+        Skill skill;
+        int baseChance;
+        Function<Client, Integer> estimateActions;
+
+        @Override
+        Double getProbability(Client client, KillCountService kcService) {
+            int level = client.getRealSkillLevel(skill);
+            return 1.0 / (baseChance - 25 * level);
```

Comment:
known shortcoming: we assume constant drop rate instead of adjusting for lower rates with lower skill levels in the past

Reply:
Fixed in https://github.com/pajlads/DinkPlugin/pull/433/commits/b98790da516d3d81fd0431782c5a225cd63bfca4 by refactoring logic from `MultiKcSource` to `MultiSource`

Reply:
technically not fixed for giant squirrel, but that's too annoying to accommodate different agility courses & skill levels over time (i.e., wontfix)